pipeline.go

```diff
@@ -0,0 +1,187 @@
+// Plugins that wish to extend the available game stages should call RegisterPipelineStageError
+// to add additional stages.
+var globalRegistry = StageRegistry{
+	StageSpawnFoodNoFood:                     RemoveFoodConstrictor,
+	StageSpawnFoodStandard:                   SpawnFoodStandard,
+	StageGameOverSoloSnake:                   GameOverSolo,
+	StageGameOverBySquad:                     GameOverSquad,
+	StageGameOverStandard:                    GameOverStandard,
+	StageHazardDamageStandard:                DamageHazardsStandard,
+	StageSpawnHazardsShrinkMap:               PopulateHazardsRoyale,
+	StageStarvationStandard:                  ReduceSnakeHealthStandard,
+	StageEliminationResurrectSquadCollisions: ResurrectSnakesSquad,
+	StageFeedSnakesStandard:                  FeedSnakesStandard,
+	StageEliminationStandard:                 EliminateSnakesStandard,
+	StageModifySnakesAlwaysGrow:              GrowSnakesConstrictor,
+	StageMovementStandard:                    MoveSnakesStandard,
+	StageMovementWrapBoundaries:              MoveSnakesWrapped,
+	StageModifySnakesShareAttributes:         ShareAttributesSquad,
+}
+
+// RegisterPipelineStage adds a stage to the registry.
+// If a stage has already been mapped it will be overwritten by the newly
+// registered function.
+func (sr StageRegistry) RegisterPipelineStage(s string, fn StageFunc) {
+	sr[s] = fn
+}
+
+// RegisterPipelineStageError adds a stage to the registry.
+// If a stage has already been mapped an error will be returned.
+func (sr StageRegistry) RegisterPipelineStageError(s string, fn StageFunc) error {
+	if _, ok := sr[s]; ok {
+		return errors.New("stage has already been registered")
+	}
+
+	sr.RegisterPipelineStage(s, fn)
+	return nil
+}
+
+// RegisterPipelineStage adds a stage to the global stage registry.
+// If a stage has already been mapped it will be overwritten by the newly
+// registered function.
+func RegisterPipelineStage(s string, fn StageFunc) {
+	globalRegistry.RegisterPipelineStage(s, fn)
+}
+
+// RegisterPipelineStageError adds a stage to the global stage registry.
+// If a stage has already been mapped an error will be returned.
+func RegisterPipelineStageError(s string, fn StageFunc) error {
+	return globalRegistry.RegisterPipelineStageError(s, fn)
+}
+
+// Pipeline is an ordered sequences of game stages which are executed to produce the
+// next game state.
+//
+// If a stage produces an error or an ended game state, the pipeline is halted at that stage.
+type Pipeline interface {
+	// Execute runs a sequence of stages and produces a next game state.
+	//
+	// If any stage produces an error or an ended game state, the pipeline
+	// immediately stops at that stage.
+	//
+	// The result is the result of the last stage that was executed.
+	//
+	Execute(*BoardState, Settings, []SnakeMove) (bool, *BoardState, error)
+	// Error can be called to check if the pipeline is in an error state.
+	Error() error
```

Comment:
Might be worth specifying that `Error()` won't return errors returned by the stages, just errors initializing the pipeline.

Reply:
Might be a reason to call it `Valid()` instead? Not sure about that.

Reply:
Hmm yeah, I was also thinking about whether something like `Valid()` is better. However, I felt that `Error()` was still fine. It's not opinionated about how the implementing pipeline treats it. It's possible that a particular implementation could return errors for other reasons. Overall, I think `Error()` is less opinionated about the implementation, because it doesn't force a validation concept and supports a lot of possible error states. I think this makes it more flexible.

Reply:
Just for context on my thinking here. Let me know if you still feel that `Valid` is a better direction and I will make it so :+1:

Reply:
Makes sense - I'm happy with `Err` or `Error` as long as it's clear that you still need to check the error from `Execute` as well.

Reply:
Thanks. I renamed it to `Err`. I think this helps to prevent overlap with the standard `Error()` method being used to create custom error types. Also made changes to the documentation to clarify things, like reinforcing the fact that the `Execute` error must always be checked and the results discarded if the error is non-nil. I think this should go without saying, but it doesn't hurt to be explicit either :)